sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -368,6 +373,98 @@ public virtual async Task<Response<UploadBlobResult>> UploadBlobAsync(Stream str
             }
         }
 
+        private async Task<ChunkedUploadResult> UploadInChunks(string location, Stream stream, int maxChunkSize, bool async = true, CancellationToken cancellationToken = default)
+        {
+            long length = stream.Length;
+
+            string digest = default;
+            ResponseWithHeaders<ContainerRegistryBlobUploadChunkHeaders> uploadChunkResult = null;
+
+            if (length < maxChunkSize)
+            {
+                digest = OciBlobDescriptor.ComputeDigest(stream);
+
+                uploadChunkResult = async ?
+                    await _blobRestClient.UploadChunkAsync(location, stream, cancellationToken: cancellationToken).ConfigureAwait(false) :
+                    _blobRestClient.UploadChunk(location, stream, cancellationToken: cancellationToken);
+
+                return new ChunkedUploadResult(digest,
+                    uploadChunkResult.Headers.Location,
+                    length);
+            }
+
+            byte[] buffer = new byte[maxChunkSize];
+            var ranges = GetRanges(length, maxChunkSize);
+            using SHA256 sha256 = SHA256.Create();
+
+            for (int i = 0; i < ranges.Count; i++)
+            {
+                var range = ranges[i];
+
+                location = uploadChunkResult?.Headers.Location ?? location;
+                var chunkLength = async ?
+                    await stream.ReadAsync(buffer, 0, (int)range.Length, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(buffer, 0, (int)range.Length);
+
+                if (i < ranges.Count - 1)
+                {
+                    sha256.TransformBlock(buffer, 0, chunkLength, buffer, 0);
+                }
+                else
+                {
+                    sha256.TransformFinalBlock(buffer, 0, chunkLength);
+                    digest = "sha256:" + OciBlobDescriptor.BytesToString(sha256.Hash);
+                }
+
+                using (Stream chunk = new MemoryStream(buffer, 0, chunkLength))
+                {
+                    uploadChunkResult = async ?
```

Comment:
>uploadChunkResult [](http://example.com/codeflow?start=20&length=17)

Given this is costly, should we consider an experience where we do not compute the digest?

This was one question I had previously as well. Shouldn't the service be doing the digest computation? I vaguely remember you mentioned that the service team explicitly asked us to support digest but I am curious do they do it and want us to support it as well or they never do it?

Reply:
For upload, we're required to send the digest in the final PUT request sent to complete the upload, after all the chunks have been sent.  The registry will also compute the digest of the blob, and will [compare the two](https://docs.docker.com/registry/spec/api/#completed-upload) to confirm that the blob hasn't been tampered with during its transfer across the wire.  So, I don't think we can avoid computing the digest for upload.